[Swift in WebKit] No WK_MAC_TBA equivalent for Swift code

rr-codes opened the pull request:
#### 56a7ae11bd6d4a8813561ffccebb23aa257d0b25
<pre>
[Swift in WebKit] No WK_MAC_TBA equivalent for Swift code
<a href="https://bugs.webkit.org/show_bug.cgi?id=278572">https://bugs.webkit.org/show_bug.cgi?id=278572</a>
<a href="https://rdar.apple.com/134584059">rdar://134584059</a>

Reviewed by Elliott Williams.

Use custom Swift availability attributes to faciliate a mechanism similar to the existing TBA macros.

* Source/WebKit/Configurations/Base.xcconfig:

Canonical link: <a href="https://commits.webkit.org/286580@main">https://commits.webkit.org/286580@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a92f585fef993558a301f15e4c5a9549998f8f2b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76331 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55363 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29233 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80843 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27603 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64505 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3657 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59854 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17974 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79398 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49755 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65546 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40194 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47153 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25925 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68280 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23367 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82298 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3703 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2423 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68082 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3857 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65518 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67384 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16817 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11349 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3651 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6458 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3674 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7103 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5432 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/Configurations/Base.xcconfig:
```diff
@@ -117,7 +117,8 @@ SUPPORTS_MACCATALYST = YES;
 SWIFT_VERSION = 5.0;
-OTHER_SWIFT_FLAGS = $(inherited) $(OTHER_SWIFT_FLAGS$(WK_MACOS_1400)) -no-warnings-as-errors -Xfrontend -experimental-spi-only-imports -Xfrontend -enable-experimental-concurrency;
+OTHER_SWIFT_FLAGS_ADDITIONS = @$(WK_WEBKITADDITIONS_HEADERS_FOLDER_PATH)/WebKit/AdditionalResources/availability-definitions.txt
+OTHER_SWIFT_FLAGS = $(inherited) $(OTHER_SWIFT_FLAGS$(WK_MACOS_1400)) -no-warnings-as-errors -Xfrontend -experimental-spi-only-imports -Xfrontend -enable-experimental-concurrency $(OTHER_SWIFT_FLAGS_ADDITIONS);
```
nits:
- Not sure "ADDITIONS" feels right, unless the goal is for the file to eventually contain arbitrary Internal-only swift flags. I think I'd name it OTHER_SWIFT_FLAGS_AVAILABILITY.
- We usually write the top-level setting first, and then the specializations below. xcconfig is not an imperative language, so we can write declarations in any order.
- Our code style is to use semicolons (though they are not required by the language).

Also, you'll need to use USE_INTERNAL_SDK here. So if you put all of this feedback together, it would look like this:

```suggestion
OTHER_SWIFT_FLAGS = $(inherited) $(OTHER_SWIFT_FLAGS$(WK_MACOS_1400)) -no-warnings-as-errors -Xfrontend -experimental-spi-only-imports -Xfrontend -enable-experimental-concurrency $(OTHER_SWIFT_FLAGS_AVAILABILITY_$(USE_INTERNAL_SDK));

OTHER_SWIFT_FLAGS_AVAILABILITY_YES = @$(WK_WEBKITADDITIONS_HEADERS_FOLDER_PATH)/WebKit/AdditionalResources/availability-definitions.txt;

```

emw-apple reviewed (commented):
Could you please also set this up for JSC?

emw-apple commented on Source/WebKit/Configurations/Base.xcconfig:
```diff
@@ -117,7 +117,8 @@ SUPPORTS_MACCATALYST = YES;
 SWIFT_VERSION = 5.0;
-OTHER_SWIFT_FLAGS = $(inherited) $(OTHER_SWIFT_FLAGS$(WK_MACOS_1400)) -no-warnings-as-errors -Xfrontend -experimental-spi-only-imports -Xfrontend -enable-experimental-concurrency;
+OTHER_SWIFT_FLAGS = $(inherited) $(OTHER_SWIFT_FLAGS$(WK_MACOS_1400)) -no-warnings-as-errors -Xfrontend -experimental-spi-only-imports -Xfrontend -enable-experimental-concurrency $(OTHER_SWIFT_FLAGS_AVAILABILITY_$(USE_INTERNAL_SDK));
+OTHER_SWIFT_FLAGS_AVAILABILITY_YES = @$(WK_WEBKITADDITIONS_HEADERS_FOLDER_PATH)/WebKit/AdditionalResources/availability-definitions.txt;
```
This isn't the right path -- clients of WebKitAdditions need to refer to its path in the SDK. So `$(WK_WEBKITADDITIONS_HEADERS_FOLDER_PATH)` will be a path ending in /usr/local/include/WebKitAdditions, and whatever comes next will be the install location of this file.

If you like what I suggested internally, then this would:
```suggestion
OTHER_SWIFT_FLAGS_AVAILABILITY_YES = @$(WK_WEBKITADDITIONS_HEADERS_FOLDER_PATH)/Scripts/availability-definitions.txt;
```

rr-codes committed:
[Swift in WebKit] No WK_MAC_TBA equivalent for Swift code
https://bugs.webkit.org/show_bug.cgi?id=278572
rdar://134584059

Reviewed by Elliott Williams.

Use custom Swift availability attributes to faciliate a mechanism similar to the existing TBA macros.

* Source/WebKit/Configurations/Base.xcconfig:

Canonical link: https://commits.webkit.org/286580@main
